File: webapp/src/stores/user_legal.ts
import { defineStore } from "pinia";
import { ref } from "vue";

interface UserLegalData {
  taxId: string;
  documentType: string;
  documentCountry: string;

  isCompany: boolean;
  companyName: string;
  companyAddress: string;
  companyCity: string;
  companyState: string;
  companyZip: string;
  companyCountry: string;
}

export const useUserLegalStore = defineStore("userLegal", () => {
  // State
  const legalData = ref<UserLegalData | null>(null);

  // Actions
  function setLegalData(data: UserLegalData) {
    legalData.value = data;
  }

  function resetLegalData() {
    legalData.value = null;
  }

  async function fetchLegalData(userId: string) {
    try {
      const response = await fetch(`/api/users/${userId}/legal`);
      if (!response.ok) throw new Error("Failed to fetch legal data");
      legalData.value = await response.json();
    } catch (error) {
      console.error("Error fetching legal data:", error);
    }
  }

  return { legalData, setLegalData, resetLegalData, fetchLegalData };
});

File: api/controllers/user_legal.go
package controllers

import (
	"github.com/bount-ing/bount.ing/api/db"
	"github.com/bount-ing/bount.ing/api/models"
	"gorm.io/gorm"
)

func CreateUserLegalData(userLegalData models.UserLegalData) error {
	if err := db.DB.Create(&userLegalData).Error; err != nil {
		return err
	}

	return nil
}

func GetUserLegalData(userID uint) (models.UserLegalData, error) {
	var userLegalData models.UserLegalData

	dbc := db.DB.Where("user_id = ?", userID).First(&userLegalData)
	if dbc.Error == gorm.ErrRecordNotFound {
		return userLegalData, nil
	} else if dbc.Error != nil {
		return userLegalData, dbc.Error
	}
	return userLegalData, dbc.Error
}

func UpdateUserLegalData(userID uint, userLegalData models.UserLegalData) error {
	var userLegalDataInDB models.UserLegalData

	dbc := db.DB.Where("user_id = ?", userID).First(&userLegalDataInDB)
	if dbc.Error == gorm.ErrRecordNotFound {
		return dbc.Error
	}

	userLegalDataInDB.TaxID = userLegalData.TaxID
	userLegalDataInDB.DocumentType = userLegalData.DocumentType
	userLegalDataInDB.DocumentCountry = userLegalData.DocumentCountry
	userLegalDataInDB.IsCompany = userLegalData.IsCompany
	userLegalDataInDB.CompanyName = userLegalData.CompanyName
	userLegalDataInDB.CompanyAddress = userLegalData.CompanyAddress
	userLegalDataInDB.CompanyCity = userLegalData.CompanyCity
	userLegalDataInDB.CompanyState = userLegalData.CompanyState
	userLegalDataInDB.CompanyZip = userLegalData.CompanyZip
	userLegalDataInDB.CompanyCountry = userLegalData.CompanyCountry

	saveResult := db.DB.Save(&userLegalDataInDB)
	if saveResult.Error != nil {
		return saveResult.Error
	}
	return nil
}
